### ISSUE TITLE
sphinx api documentation

### ISSUE THREAD

**Issue by AlbinaMuzafarova** [pik-copan/pycopancore#159]
## Description

make the sphinx-generated api documentation look better, have a nice table of contents, show the package tree in a sensitive way, show the model component docstrings in the right place, etc.

## Todo

**Comment by AlbinaMuzafarova** [comment]
@leander-j @mensch72 since I am experiencing troubles generating the documentation from master, is this what you see when you generate the documentation currently?

![image](https://user-images.githubusercontent.com/36408342/146199246-8943a9d3-7ce6-4a06-893e-02fce13ef4c7.png)

The `Index` is empty and `Module Index` for API reference not found

**Comment by leander-j** [comment]
`index.html` looks similar:
![sphinx_master_doc](https://user-images.githubusercontent.com/64526813/146209726-41bdcbf4-aaa6-41c1-9f56-7dda560cb70d.png)

The `Index` and `Module Index` are not empty though.

I noticed that on other branches (e.g. imitation_component/issue_152) at least the TOC structure is different:
![sphinx_doc](https://user-images.githubusercontent.com/64526813/146210307-92933e8f-3db2-4374-8c33-cb847133c03d.png)

**Comment by mensch72** [comment]
in my case both index and module index are populated. otherwise it looks the same. i'll send you a zip of the html folder.

Am 15. Dezember 2021 15:10:15 MEZ schrieb AlbinaMuzafarova ***@***.***>:
***@***.*** @mensch72 since I am experiencing troubles generating the
>documentation from master, is this what you see when you generate the
>documentation currently?
>
>![image](https://user-images.githubusercontent.com/36408342/146199246-8943a9d3-7ce6-4a06-893e-02fce13ef4c7.png)
>
>The `Index` is empty and `Module Index` for API reference not found
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pik-copan/pycopancore/issues/159#issuecomment-994826662

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by AlbinaMuzafarova** [comment]
thank you! I'll compare the received html with mine for deviations and would be also curious to check the config file for the branch mentioned by @leander-j

**Comment by leander-j** [comment]
The screenshots were from the [issue_152 branch](https://github.com/pik-copan/pycopancore/tree/issue_152), the html generated there is
[html.zip](https://github.com/pik-copan/pycopancore/files/7720861/html.zip)

**Comment by AlbinaMuzafarova** [comment]
@leander-j i've changed the depth of the table for `issue_152` branch. It should be 1 now, same as for master. 

@mensch72 @leander-j I think i'm generating the index and module index properly now, although there are still many errors in the terminal. Could you let me know if this is what you see (I've changed the theme, so please don't consider the difference in colors):

![image](https://user-images.githubusercontent.com/36408342/146675932-f959ad98-a38d-4d2d-86ef-ef0c918cde5b.png)

![image](https://user-images.githubusercontent.com/36408342/146675967-f4d766a4-e54c-4c26-9b7c-72a781b882c2.png)

**Comment by leander-j** [comment]
Yes, it does look like that, just with way more modules/items. In fact so many that the Index page hides them
![sphinx_index](https://user-images.githubusercontent.com/64526813/146815130-2bb51c9a-dfe9-437b-bd0d-ce2674cf8c8f.png)

**Comment by AlbinaMuzafarova** [comment]
I've pushed first changes to `doc/api` branch. If someone could try generating the documentation locally on that branch, that'd be great. 

Expected output: ToC with API reference which is populated according to the description

I've adapted the theme to Book theme because it handles presentation and navigation between pages/headings smoother than alabaster. To install, use `pip install sphinx-book-theme`

I suspect that due to me running Windows, sphinx is not working as smoothly as for the rest of the team. To double-check my assumptions, i'd hearing whether these parts of the `API reference>Models` page are filled:

![image](https://user-images.githubusercontent.com/36408342/151373515-3a237aec-bd8e-4956-88b9-5af8c4bd79ac.png)

**Comment by mensch72** [comment]
Hi Albina,

overall this looks much nicer than before! Thank you!!

I do have a number of fast observations:

- In the master data model, the Variables do no longer show up: file:///home/heitzig/Nextcloud/git/pycopancore/docs/_build/html/api/pycopancore/pycopancore.data_model.master_data_model.html#module-pycopancore.data_model.master_data_model.environment Before, each defined Variable was listed with its metadata. (Currently, the dimensions and units are shown, file:///home/heitzig/Nextcloud/git/pycopancore/docs/_build/html/api/pycopancore/pycopancore.data_model.master_data_model.html#module-pycopancore.data_model.master_data_model.dimensions_and_units and I imagine it similarly for Variables). 
- It is somewhat unclear to me, where the section titles / headlines used in the APIdoc are coming from. E.g., in file:///home/heitzig/Nextcloud/git/pycopancore/docs/_build/html/api/pycopancore/pycopancore.data_model.master_data_model.html#module-pycopancore.data_model.master_data_model.dimensions_and_units where does the headline "Dimensions and units" come from, I can't find it in the source. I would suggest that sections for classes should always show the class name (rather than some part of the docstring), e.g. "DimensionsAndUnits" instead of ""Dimensions and units".
- We should find a way to distinguish "public" methods meant to be used by model users (like "run" in "Runner") from "private" methods only used internally in the framework (like "apply_explicits" in "Runner", see file:///home/heitzig/Nextcloud/git/pycopancore/docs/_build/html/api/pycopancore/pycopancore.runners.html#module-pycopancore.runners.runner). 
- Parameters of class constructors sometimes do not seem to appear in the doc, e.g. in "Runner", the parameter descriptions of "model" and "termination_calls" in the docstring don't seem to appear.
- For model components, it is still not clear to me, where the natural place for introductory text that will then shows up at the top of the page would be. Either of the following places could seem natural, but I don't know how to make Sphinx show them at the top of the model component page:
  - a dedicated README.md file in the subpackage folder
  - the module docstring in "__init__.py", "model.py" or "interface.py"
  - the class doctring of "Model" in model.py

**Comment by AlbinaMuzafarova** [comment]
thank you Jobst for your first feedback! I'll gradually walk my way from the bottom. 
> For model components, it is still not clear to me, where the natural place for introductory text that will then shows up at the top of the page would be. 

Currently, sphinx looks at `_init_.py` file and present the docstring on the top of the model component/any other page. So I would suggest we keep using that location for general descriptions.

**Comment by AlbinaMuzafarova** [comment]
> It is somewhat unclear to me, where the section titles / headlines used in the APIdoc are coming from

The headings were generated by sphinx first as paths and seemed unreadable to me, e.g. `pycopancore/pycopancore.data_model.dimensions_and_units`. So I've modified them manually for better user experience. I can adjust the subtitles to reflect the exact naming of the package/class in the next commit.

**Comment by AlbinaMuzafarova** [comment]
> We should find a way to distinguish "public" methods meant to be used by model users (like "run" in "Runner") from "private" methods only used internally in the framework

I think sphinx has a flag to hide private methods. Would that be a desired solution here or were you thinking of finding just a visual distinction?

**Comment by mensch72** [comment]
that would be perfect

Am 31. Januar 2022 17:58:45 MEZ schrieb AlbinaMuzafarova ***@***.***>:
>> We should find a way to distinguish "public" methods meant to be used
>by model users (like "run" in "Runner") from "private" methods only
>used internally in the framework
>
>I think sphinx has a flag to hide private methods. Would that be a
>desired solution here or were you thinking of finding just a visual
>distinction?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/pik-copan/pycopancore/issues/159#issuecomment-1025996120
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.
